oracle/internal/dkg/dkg_executor.go

```diff
@@ -254,45 +285,52 @@ func (e *Executor) executeR2(dkg *coordinator.DKG, validatorIdx uint16) bool {
 			return false
 		}
 
-		e.artifacts.r2 = &Round2Result{
-			pkgs:   r2Packages,
-			secret: NewSecret(r2SecretPtr),
-		}
-	}
+		// Convert R2 packages to bytes
+		var r2packagesBatched []byte
+		tmpBuf := make([]byte, 2)
+		for toIdentificator, r2pkg := range r2Packages {
+			// Serialize validator idx
+			toValidatorIdx := helpers.FrostToValidatorIdx(toIdentificator)
+			binary.BigEndian.PutUint16(tmpBuf, toValidatorIdx)
+			r2packagesBatched = append(r2packagesBatched, tmpBuf...)
+
+			// Encrypt r2pkg
+			r2PublicKeyX25519, ok := r2PublicKeysX25519[toValidatorIdx]
+			if !ok {
+				e.logError(dkg, fmt.Sprintf("No X25519 public key was found for Oracle {%d}", toValidatorIdx), nil)
+				return false
+			}
 
-	e.logMessage(dkg, "sending r2 packages...")
-	withErrors := false
-	// Get r2 packages that are already sent to coordinator.
-	// But only from this oracle to others
-	// Go through all r2 packages generated locally
-	for toIdentificator, r2pkg := range e.artifacts.r2.pkgs {
-		// Check if oracle has already sent this package to coordinator
-
-		toIdx := helpers.FrostToValidatorIdx(toIdentificator)
-		sentPackages, foundToPackages := dkg.GetR2PackagesTo(toIdx)
-
-		if foundToPackages {
-			_, foundFromMePackage := sentPackages[validatorIdx]
-			if foundFromMePackage {
-				continue
+			r2pkgEncrypted, err := Encrypt(r2pkg.ToBytes(), &e.artifacts.r1.r2PrivateX25519, (*[32]byte)(r2PublicKeyX25519[:32]))
+			if err != nil {
+				e.logError(dkg, fmt.Sprintf("Failed to encrypt R2 packages for Oracle {%d}", toValidatorIdx), err)
+				return false
 			}
+
+			// Serialize r2pkgEncrypted
+			binary.BigEndian.PutUint16(tmpBuf, uint16(len(r2pkgEncrypted)))
+			r2packagesBatched = append(r2packagesBatched, tmpBuf...)
+			r2packagesBatched = append(r2packagesBatched, r2pkgEncrypted...)
```

Comment:
I think it's better to make a separate function for encrypted serialization.
Because you have DecryptR2Packages but don't have EncryptR2Packages

Reply:
And why don't you use map for storing encrypted packages? map[uint16][]byte ?